k8s/operators/src/pool/diskpool/client.rs

```diff
@@ -35,33 +71,165 @@ pub(crate) async fn create(
+            serde_json::to_string_pretty(&new_crd).unwrap()
+        );
+        update_stored_version(k8s, crd_name, new_version).await?;
+        if let Ok(modified_crd) = crd_api.get(crd_name).await {
+            new_crd.metadata.resource_version = modified_crd.resource_version();
+            let pp = PostParams::default();
+            crd_api.replace(crd_name, &pp, &new_crd).await?;
+        }
+    } else {
+        info!(
+            "Creating CRD: {}",
+            serde_json::to_string_pretty(&new_crd).unwrap()
+        );
+        let pp = PostParams::default();
+        if let Err(e) = crd_api.create(&pp, &new_crd).await {
+            error!("Failed to create v1beta1 CRD: {:?}", e);
+        }
+    }
+    Ok(())
+}
+
+/// Lists existing v1alpha1 CR in cluster and replaces them with v1beta1 CR.
+/// This ensures that there is no v1alpha1 stored objects in cluster.
+pub(crate) async fn migrate_to_v1beta1(
+    k8s: Client,
+    ns: &str,
+    pagination_limit: u32,
+) -> Result<(), Error> {
+    if let Ok(mut v1_alpha1_pools) = list_existing_cr(&k8s, ns, pagination_limit).await {
+        for dsp in v1_alpha1_pools.iter_mut() {
+            if let Some(res_ver) = dsp.resource_version() {
+                let name = dsp.name_any();
+                let node = dsp.spec.node();
+                let disk = dsp.spec.disks();
+                replace_with_v1beta1(
+                    &k8s,
+                    &name,
+                    ns,
+                    Some(res_ver.clone()),
+                    DiskPoolSpec::new(node, disk),
+                )
+                .await?;
+                info!(crd = ?dsp.name_any(), "CR creation successful");
+            } else {
+                return Err(Error::CrdFieldMissing {
+                    name: dsp.name_any(),
+                    field: "resource_version".to_string(),
+                });
+            }
+        }
+    } else {
+        return Err(Error::Generic {
+            message: "Error in listing v1alpha1 CR".to_string(),
+        });
+    };
+    Ok(())
+}
+
+/// Updates stored version in CRD status to the new version passed as arg,
+/// Please ensure that older versions are served:false, stored:false before calling this method,
+/// This would allow us to remove older schema from CRD versions.
+async fn update_stored_version(
+    k8s: &Client,
+    crd_name: &str,
+    new_version: &str,
+) -> Result<(), Error> {
+    let crd_api: Api<CustomResourceDefinition> = Api::all(k8s.clone());
+    if let Ok(mut crd) = crd_api.get_status(crd_name).await {
+        let param = PatchParams::apply("status_patch").force();
+        if let Some(mut status) = crd.status.as_mut() {
+            status.stored_versions = Some(vec![new_version.to_string()]);
+        } else {
+            return Err(Error::CrdFieldMissing {
+                name: crd_name.to_string(),
+                field: "status".to_string(),
+            });
+        }
+        crd.metadata.managed_fields = None;
```

Comment:
what does setting this do?

Reply:
What ive seen is whiling patching the existing object (which we get from kubeapi server as mutable) it expects managed_fields to be Nil. Error was specific in patch api. I printed with setting it to None it was showing some metadata in it. With this patch worked.

Reply:
Should it be reverted back afterwards?

Reply:
I dont know i think it is set by k8s. I can check if it gets set again after the patch..